# Yandex.Music

# Contents <a id='back'></a>

* [Introduction](#intro)
* [Stage 1. Data overview](#data_review)
    * [Conclusions](#data_review_conclusions)
* [Stage 2. Data preprocessing](#data_preprocessing)
    * [2.1 Header style](#header_style)
    * [2.2 Missing values](#missing_values)
    * [2.3 Duplicates](#duplicates)
    * [2.4 Conclusions](#data_preprocessing_conclusions)
* [Stage 3. Testing the hypotheses](#hypotheses)
    * [3.1 Hypothesis 1: user activity in the two cities](#activity)
    * [3.2 Hypothesis 2: music preferences on Monday and Friday](#week)
    * [3.3 Hypothesis 3: genre preferences in Springfield and Shelbyville](#genre)
* [Findings](#end)

## Introduction <a id='intro'></a>
Whenever we're doing research, we need to formulate hypotheses that we can then test. Sometimes we accept these hypotheses; other times, we reject them. To make the right decisions, a business must be able to understand whether or not it's making the right assumptions.

In this project, you'll compare the music preferences of the cities of Springfield and Shelbyville. You'll study real Yandex.Music data to test the hypotheses below and compare user behavior for these two cities.

### Goal: 
Test three hypotheses:
1. User activity differs depending on the day of the week and from city to city. 
2. On Monday mornings, Springfield and Shelbyville residents listen to different genres. This is also true for Friday evenings. 
3. Springfield and Shelbyville listeners have different preferences. In Springfield, they prefer pop, while Shelbyville has more rap fans.

### Stages 
Data on user behavior is stored in the file `/datasets/music_project_en.csv`. There is no information about the quality of the data, so you will need to explore it before testing the hypotheses. 

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Testing the hypotheses
 
[Back to Contents](#back)

## Stage 1. Data overview <a id='data_review'></a>

Open the data on Yandex.Music and explore it.

In [1]:
# importing pandas
import pandas as pd

Read the file `music_project_en.csv` from the `/datasets/` folder and save it in the `df` variable:

In [2]:
# reading the file and storing it to df / absolute file path
# df = pd.read_csv('/datasets/music_project_en.csv')

# summary statistics
# df.describe()

df2 = pd.read_csv('/Users/dani/Data Science/TripleTen Projects/Project Data/Basic Python/movies_and_shows.csv')

display(df2)

name                Character      r0le        TITLE  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
85574    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
85575  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
85576        In??s Prieto                    Fanny     ACTOR      Lokillo   
85577        Isabel Gaona                   Cacica     ACTOR      Lokillo   
85578      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            Type  release Year              genres  imdb sc0re  imdb v0tes  
0          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
1          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
2          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
3          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
4          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
...          ...           ...                 ...         ...         ...  
85574  the movie          2021          ['comedy']         3.8        68.0  
85575  the movie          2021          ['comedy']         3.8        68.0  
85576  the movie          2021          ['comedy']         3.8        68.0  
85577      MOVIE          2021          ['comedy']         3.8        68.0  
85578  the movie          2021          ['comedy']         3.8        68.0  

[85579 rows x 9 columns]

Print the first 10 table rows:

In [262]:
# obtaining the first 10 rows from the df table
df.head(10)

# obtaining the last 10 rows from the df table in case it's needed for further analysis
# df.tail(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtaining the general information about the table with one command:

In [263]:
# obtaining general information about the data in df
df.info()

# expected issues with the data under: Track/artist/genre
# expected issues with column names: userID and City

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


**`Comments`**

1. expected issues with the data under: Track/artist/genre

2. expected issues with column names: userID and City

The table contains seven columns. They all store the same data type: `object`.

According to the documentation:
- `'userID'` — user identifier
- `'Track'` — track title
- `'artist'` — artist's name
- `'genre'`
- `'City'` — user's city
- `'time'` — the exact time the track was played
- `'Day'` — day of the week

We can see three issues with style in the column names:
1. Some names are uppercase, some are lowercase.
2. There are spaces in some names.
3. Column names (where applicable) are missing the snake_case syntax.

The number of column values is different. This means the data contains missing values.

### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data on a track that was played. Some columns describe the track itself: its title, artist and genre. The rest convey information about the user: the city they come from, the time they played the track. 

It's clear that the data is sufficient to test the hypotheses. However, there are missing values.

To move forward, we need to preprocess the data.

[Back to Contents](#back)

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

### Header style <a id='header_style'></a>
Print the column header:

In [264]:
# the list of column names in the df table

print(df.columns)

# syntax and spacing issues become visible / all string types

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Change column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Delete spaces

In [265]:
# renaming columns for suitable presentation of the data
# to be more descriptive, can also do 'track_title' and 'user_city'

df = df.rename(
    columns = {
        '  userID': 'user_id',
        'Track': 'track',
        '  City  ': 'city',
        'Day': 'day',
    }
)

**`Comments`**

To be more descriptive above, can also do 'track_title' and 'user_city'


Check the result. Print the names of the columns once more:

In [266]:
# checking result: the list of column names

print(df.columns)


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Back to Contents](#back)

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, use two `pandas` methods:

In [267]:
# calculating missing values combining the isna() and sum() methods

print(df.isna().sum())

# analysis shows there are missing values under 3 column names, matching the variance from the .info() call from earlier

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


**`Comments`**

Analysis shows there are missing values under 3 column names, matching the variance from the .info() call from earlier

Not all missing values affect the research. For instance, the missing values in `track` and `artist` are not critical. You can simply replace them with clear markers.

But missing values in `'genre'` can affect the comparison of music preferences in Springfield and Shelbyville. In real life, it would be useful to learn the reasons why the data is missing and try to make up for them. But we do not have that opportunity in this project. So you will have to:
* Fill in these missing values with markers
* Evaluate how much the missing values may affect your computations

Replace the missing values in `'track'`, `'artist'`, and `'genre'` with the string `'unknown'`. To do this, create the `columns_to_replace` list, loop over it with `for`, and replace the missing values in each of the columns:

In [303]:
# looping over column names and replacing missing values with 'unknown' 
# using indexing and the fillna() method - understand this would be a hassle if there were more than 3 columns to replace with missing values
    #*for the reviewer - I initially toyed with indexing and now I can't get it to revert back to original to attempt the looping (should have saved a copy)#

# df['track'] = df['track'].fillna('unknown')
# df['artist'] = df['artist'].fillna('unknown')
# df['genre'] = df['genre'].fillna('unknown')

track_name = 'track'
artist_name = 'artist'
genre_name = 'genre'

columns_to_replace = [
    df[track_name],
    df[artist_name], 
    df[genre_name],
]

# print(columns_to_replace)

for col in range(len(columns_to_replace)):
    columns_to_replace[col].fillna('unknown', inplace=True)
#    print(columns_to_replace[col])

# print(df[artist_name].isna().sum())


# df.info() # check

Make sure the table contains no more missing values. Count the missing values again.

In [304]:
# counting missing values

print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Back to Contents](#back)

### Duplicates <a id='duplicates'></a>
Find the number of obvious duplicates in the table using one command:

In [271]:
# counting clear duplicates

df.duplicated().sum()

# using logical indexing to produce a table of duplicates
# duplicated_df = df[df.duplicated()]
# print(duplicated_df)

3826

**`Comments`**

3826 rows with duplicates, matches the count from the earlier code

Call the `pandas` method for getting rid of obvious duplicates:

In [272]:
# removing obvious duplicates using the drop_duplicates() method

df = df.drop_duplicates()

# note to check indices, reset_index()

Count obvious duplicates once more to make sure you have removed all of them:

In [273]:
# checking for duplicates

df.duplicated().sum()

0

**`Comments`**


0 value shows that all duplicates have been removed

Now get rid of implicit duplicates in the `genre` column. For example, the name of a genre can be written in different ways. Such errors will also affect the result.

Print a list of unique genre names, sorted in alphabetical order. To do so:
* Retrieve the intended DataFrame column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [274]:
# viewing unique genre names using the unique() method, returning a list

# creating a variable to store the DataFrame column/Series
unique_list = df['genre']

# creating a variable to store retrieve column and sort in alphabetical order
sorted_genre = unique_list.sort_values(ascending=True)

# print(sorted_genre) # check

# printing the sorted list with all unique values
print(sorted_genre.unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

**`Comments`**


Could have also combined the above into one code block:
df['genre'].sort_values(ascending=True).unique()

Look through the list to find implicit duplicates of the genre `hiphop`. These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `hip`
* `hop`
* `hip-hop`

To get rid of them, declare the function `replace_wrong_genres()` with two parameters: 
* `wrong_genres=` — the list of duplicates
* `correct_genre=` — the string with the correct value

The function should correct the names in the `'genre'` column from the `df` table, i.e. replace each value from the `wrong_genres` list with the value in `correct_genre`.

In [275]:
# function for replacing implicit duplicates where we pass two parameters the wrong and the correct genres

def replace_wrong_genres(wrong_genres, correct_genre):
    # looping over the list of duplicates
    for wrong_genre in wrong_genres:
        # using indexing to access genre column and correct the names
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)


Call `replace_wrong_genres()` and pass it arguments so that it clears implicit duplcates (`hip`, `hop`, and `hip-hop`) and replaces them with `hiphop`:

In [276]:
# removing implicit duplicates via two variables and a function call

# creating two variables which will store 1) the incorrect spelling and 2) the correct spelling of target string
implicit_duplicates = ('hip', 'hop', 'hip-hop')
correct_name = 'hiphop'

# calling the function with variables as arguments
replace_wrong_genres(implicit_duplicates, correct_name)

# print(df)

Make sure the duplicate names were removed. Print the list of unique values from the `'genre'` column:

In [277]:
# checking for implicit duplicates and validating that only the correct name remains in 'hiphop'

print(df['genre'].sort_values(ascending=True).unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

[Back to Contents](#back)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Obvious and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All missing values have been replaced with `'unknown'`. But we still have to see whether the missing values in `'genre'` will affect our calculations.

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to testing hypotheses. 

[Back to Contents](#back)

## Stage 3. Testing hypotheses <a id='hypotheses'></a>

### Hypothesis 1: comparing user behavior in two cities <a id='activity'></a>

According to the first hypothesis, users from Springfield and Shelbyville listen to music differently. Test this using the data on three days of the week: Monday, Wednesday, and Friday.

* Divide the users into groups by city.
* Compare how many tracks each group played on Monday, Wednesday, and Friday.


For the sake of practice, perform each computation separately. 

Evaluate user activity in each city. Group the data by city and find the number of songs played in each group.



In [278]:
# Counting up the tracks played in each city using the groupby() method

df.groupby('city').count()

user_id  track  artist  genre   time    day
city                                                    
Shelbyville    18512  18512   18512  18512  18512  18512
Springfield    42741  42741   42741  42741  42741  42741

**`Comments`**


Higher track count in Springfield 


In [279]:
df.groupby('city')['user_id'].count()

city
Shelbyville    18512
Springfield    42741
Name: user_id, dtype: int64

Springfield has more tracks played than Shelbyville. But that does not imply that citizens of Springfield listen to music more often. This city is simply bigger, and there are more users.

Now group the data by day of the week and find the number of tracks played on Monday, Wednesday, and Friday.


In [280]:
# Calculating tracks played on each of the three days

df.groupby('day').count()

# higher track count on Fridays

user_id  track  artist  genre   city   time
day                                                   
Friday       21840  21840   21840  21840  21840  21840
Monday       21354  21354   21354  21354  21354  21354
Wednesday    18059  18059   18059  18059  18059  18059

**`Comments`**


Higher track count on Fridays


In [281]:
df.groupby('day')[['user_id']].count()

user_id
day               
Friday       21840
Monday       21354
Wednesday    18059

Wednesday is the quietest day overall. But if we consider the two cities separately, we might come to a different conclusion.

You have seen how grouping by city or day works. Now write a function that will group by both.

Create the `number_tracks()` function to calculate the number of songs played for a given day and city. It will require two parameters:
* day of the week
* name of the city

In the function, use a variable to store the rows from the original table, where:
  * `'day'` column value is equal to the `day` parameter
  * `'city'` column value is equal to the `city` parameter

Apply consecutive filtering with logical indexing.

Then calculate the `'user_id'` column values in the resulting table. Store the result to a new variable. Return this variable from the function.

In [282]:
# <creating the function number_tracks()>
# We'll declare a function with two parameters: day=, city=.
def number_tracks(day, city):
# Let the track_list variable store the df rows where
# the value in the 'day' column is equal to the day= parameter and, at the same time, 
# the value in the 'city' column is equal to the city= parameter (apply consecutive filtering 
# with logical indexing).
    track_list = df[(df['day'] == day) & (df['city'] == city)]
# Let the track_list_count variable store the number of 'user_id' column values in track_list
# (found with the count() method).
    track_list_count = track_list['user_id'].count()
# Let the function return a number: the value of track_list_count.
    return track_list_count

# The function counts tracked played for a certain city and day.
# It first retrieves the rows with the intended day from the table,
# then filters out the rows with the intended city from the result,
# then finds the number of 'user_id' values in the filtered table,
# then returns that number.

# To see what it returns, wrap the function call in print().

# print(number_tracks)


Call `number_tracks()` six times, changing the parameter values, so that you retrieve the data on both cities for each of the three days.

In [283]:
# the number of songs played in Springfield on Monday

print(number_tracks('Monday','Springfield'))

15740


In [284]:
# the number of songs played in Shelbyville on Monday

print(number_tracks('Monday','Shelbyville'))

5614


In [285]:
# the number of songs played in Springfield on Wednesday

print(number_tracks('Wednesday','Springfield'))

11056


In [286]:
# the number of songs played in Shelbyville on Wednesday

print(number_tracks('Wednesday','Shelbyville'))

7003


In [287]:
# the number of songs played in Springfield on Friday

print(number_tracks('Friday','Springfield'))

15945


In [288]:
# the number of songs played in Shelbyville on Friday

print(number_tracks('Friday','Shelbyville'))

5895


Use `pd.DataFrame` to create a table, where
* Column names are: `['city', 'monday', 'wednesday', 'friday']`
* The data is the results you got from `number_tracks()`

In [289]:
# table with results, passing data and column names to the DataFrame call

# creating a variable storing the data from the 6 calls above
table_data = [
    ['Shelbyville', 5614,7003,5895],
    ['Springfield', 15740, 11056, 15945]
]

# creating a variable storing the respective columns for the new data structure we've created
table_columns = ['city', 'monday', 'wednesday', 'friday']

new_table = pd.DataFrame(data= table_data, columns = table_columns)

# checking results
print(new_table)

          city  monday  wednesday  friday
0  Shelbyville    5614       7003    5895
1  Springfield   15740      11056   15945


**Conclusions**

The data reveals differences in user behavior:

- In Springfield, the number of songs played peaks on Mondays and Fridays, while on Wednesday there is a decrease in activity.
- In Shelbyville, on the contrary, users listen to music more on Wednesday. User activity on Monday and Friday is smaller.

So the first hypothesis seems to be correct.

[Back to Contents](#back)

### Hypothesis 2: music at the beginning and end of the week <a id='week'></a>

According to the second hypothesis, on Monday morning and Friday night, citizens of Springfield listen to genres that differ from ones users from Shelbyville enjoy.

Get tables (make sure that the name of your combined table matches the DataFrame given in the two code blocks below):
* For Springfield — `spr_general`
* For Shelbyville — `shel_general`

In [290]:
# create the spr_general table from the df rows, 
# where the value in the 'city' column is 'Springfield'

spr_general = df[df['city'] == 'Springfield']

# check
# print(spr_general)


In [291]:
# create the shel_general from the df rows,
# where the value in the 'city' column is 'Shelbyville'

shel_general = df[df['city'] == 'Shelbyville']

# check
# print(shel_general)


Write the `genre_weekday()` function with four parameters:
* A table for data (`df`)
* The day of the week (`day`)
* The first timestamp, in 'hh:mm' format (`time1`)
* The last timestamp, in 'hh:mm' format (`time2`)

The function should return info on the 15 most popular genres on a given day within the period between the two timestamps.

In [292]:
# 1) Let the genre_df variable store the rows that meet several conditions:
#    - the value in the 'day' column is equal to the value of the day= argument
#    - the value in the 'time' column is greater than the value of the time1= argument
#    - the value in the 'time' column is smaller than the value of the time2= argument
#    Use consecutive filtering with logical indexing.

# genre_df = df[df['day'] == day & df['time'] >= time1 & df['time'] <= time2]

# 2) Group genre_df by the 'genre' column, take one of its columns, 
#    and use the count() method to find the number of entries for each of 
#    the represented genres; store the resulting Series to the
#    genre_df_count variable

# genre_def_count = genre_df.groupby('genre')['genre'].count()

# 3) Sort genre_df_count in descending order of frequency and store the result
#    to the genre_df_sorted variable

# genre_df_sorted = genre_df_count.sort_values(ascending=False)

# 4) Return a Series object with the first 15 genre_df_sorted value - the 15 most
#    popular genres (on a given day, within a certain timeframe)

# Write your function here

def genre_weekday(df, day, time1, time2):

    # consecutive filtering
    # Create the variable genre_df which will store only those df rows where the day is equal to day=
    # genre_df = df[df['day'] == day]

    # filter again so that genre_df will store only those rows where the time is smaller than time2=
    # genre_df = df[df['time'] <= time2]

    # filter once more so that genre_df will store only rows where the time is greater than time1=
    # genre_df = df[df['time'] >= time1]
    
    ## combining the filters 
    genre_df = df[(df['day'] == day) & (df['time'] >= time1) & (df['time'] <= time2)]
    
    # group the filtered DataFrame by the column with the names of genres, take the genre column, and find the number of rows for each genre with the count() method
    genre_df_count = genre_df.groupby('genre')['genre'].count()

    # sort the result in descending order (so that the most popular genres come first in the Series object)
    genre_df_sorted = genre_df_count.sort_values(ascending=False)

    # we will return the Series object storing the 15 most popular genres on a given day in a given timeframe
    return genre_df_sorted[:15]

# check
# print(genre_weekday(spr_general, 'Monday', '07:00', '11:00'))

# check
# print(genre_weekday(shel_general, 'Monday', '07:00', '11:00'))

# check
# print(genre_weekday(spr_general, 'Friday', '17:00', '23:00'))

# check
# print(genre_weekday(shel_general, 'Friday', '17:00', '23:00'))

Compare the results of the `genre_weekday()` function for Springfield and Shelbyville on Monday morning (from 7AM to 11AM) and on Friday evening (from 17:00 to 23:00):

In [293]:
# calling the function for Monday morning in Springfield (use spr_general instead of the df table)

genre_weekday(spr_general, 'Monday', '07:00', '11:00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
classical      157
metal          120
jazz           100
folk            97
soundtrack      95
Name: genre, dtype: int64

In [294]:
# calling the function for Monday morning in Shelbyville (use shel_general instead of the df table)

genre_weekday(shel_general, 'Monday', '07:00', '11:00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
world           36
rap             32
soundtrack      31
rnb             27
metal           27
Name: genre, dtype: int64

In [295]:
# calling the function for Friday evening in Springfield

genre_weekday(spr_general, 'Friday', '17:00', '23:00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
jazz           111
unknown        110
soundtrack     105
rnb             90
metal           88
Name: genre, dtype: int64

In [296]:
# calling the function for Friday evening in Shelbyville

genre_weekday(shel_general, 'Friday', '17:00', '23:00')

genre
pop            256
rock           216
electronic     216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
unknown         47
ruspop          47
soundtrack      40
metal           39
rap             36
Name: genre, dtype: int64

**Conclusion**

Having compared the top 15 genres on Monday morning, we can draw the following conclusions:

1. Users from Springfield and Shelbyville listen to similar music. The top five genres are the same, only rock and electronic have switched places.

2. In Springfield, the number of missing values turned out to be so big that the value `'unknown'` came in 10th. This means that missing values make up a considerable portion of the data, which may be a basis for questioning the reliability of our conclusions.

For Friday evening, the situation is similar. Individual genres vary somewhat, but on the whole, the top 15 is similar for the two cities.

Thus, the second hypothesis has been partially proven true:
* Users listen to similar music at the beginning and end of the week.
* There is no major difference between Springfield and Shelbyville. In both cities, pop is the most popular genre.

However, the number of missing values makes this result questionable. In Springfield, there are so many that they affect our top 15. Were we not missing these values, things might look different.

[Back to Contents](#back)

### Hypothesis 3: genre preferences in Springfield and Shelbyville <a id='genre'></a>

Hypothesis: Shelbyville loves rap music. Springfield's citizens are more into pop.

Group the `spr_general` table by genre and find the number of songs played for each genre with the `count()` method. Then sort the result in descending order and store it to `spr_genres`.

In [297]:
# on one line: group the spr_general table by the 'genre' column, 
# count the 'genre' values with count() in the grouping, 
# sort the resulting Series in descending order, and store it to spr_genres

spr_genres = spr_general.groupby('genre')['genre'].count().sort_values(ascending=False)

# check
# print(spr_genres)

Print the first 10 rows from `spr_genres`:

In [299]:

spr_genres.to_frame().head(10)

genre
genre             
pop           5892
dance         4435
rock          3965
electronic    3786
hiphop        2096
classical     1616
world         1432
alternative   1379
ruspop        1372
rusrap        1161

Now do the same with the data on Shelbyville.

Group the `shel_general` table by genre and find the number of songs played for each genre. Then sort the result in descending order and store it to the `shel_genres` table:


In [300]:
# on one line: group the shel_general table by the 'genre' column, 
# count the 'genre' values in the grouping with count(), 
# sort the resulting Series in descending order and store it to shel_genres

shel_genres = shel_general.groupby('genre')['genre'].count().sort_values(ascending=False)

# check
# print(shel_genres)

Print the first 10 rows of `shel_genres`:

In [301]:
# printing the first 10 rows from shel_genres

print(shel_genres.head(10))

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64


**Conclusion**

The hypothesis has been partially proven true:
* Pop music is the most popular genre in Springfield, as expected.
* However, pop music turned out to be equally popular in Springfield and Shelbyville, and rap wasn't in the top 5 for either city.


[Back to Contents](#back)

# Findings <a id='end'></a>

We have tested the following three hypotheses:

1. User activity differs depending on the day of the week and from city to city. 
2. On Monday mornings, Springfield and Shelbyville residents listen to different genres. This is also true for Friday evenings. 
3. Springfield and Shelbyville listeners have different preferences. In both Springfield and Shelbyville, they prefer pop.

After analyzing the data, we concluded:

1. User activity in Springfield and Shelbyville depends on the day of the week, though the cities vary in different ways. 

The first hypothesis is fully accepted.

2. Musical preferences do not vary significantly over the course of the week in both Springfield and Shelbyville. We can see small differences in order on Mondays, but:
* In Springfield and Shelbyville, people listen to pop music most.

So we can't accept this hypothesis. We must also keep in mind that the result could have been different if not for the missing values.

3. It turns out that the musical preferences of users from Springfield and Shelbyville are quite similar.

The third hypothesis is rejected. If there is any difference in preferences, it cannot be seen from this data.

### Note 
In real projects, research involves statistical hypothesis testing, which is more precise and more quantitative. Also note that you cannot always draw conclusions about an entire city based on the data from just one source.

You will study hypothesis testing in the sprint on statistical data analysis.

[Back to Contents](#back)